sklearn/semi_supervised/tests/test_self_training.py

```diff
@@ -0,0 +1,189 @@
+import numpy as np
+import pytest
+from sklearn.utils.testing import assert_array_equal
+from sklearn.utils.testing import assert_raise_message
+from sklearn.utils.testing import assert_warns_message
+from sklearn.exceptions import NotFittedError
+
+from sklearn.semi_supervised import SelfTrainingClassifier
+from sklearn.dummy import DummyClassifier
+from sklearn.neighbors import KNeighborsClassifier
+from sklearn.tree import DecisionTreeClassifier
+from sklearn.svm import SVC
+from sklearn.model_selection import train_test_split
+from sklearn.datasets import load_iris
+from sklearn.preprocessing import LabelBinarizer
+from sklearn.metrics import accuracy_score
+
+# Author: Oliver Rausch <rauscho@ethz.ch>
+# License: BSD 3 clause
+
+# load the iris dataset and randomly permute it
+iris = load_iris()
+X_train, X_test, y_train, y_test = train_test_split(iris.data,
+                                                    iris.target,
+                                                    random_state=0)
+
+n_labeled_samples = 50
+
+y_train_missing_labels = y_train.copy()
+y_train_missing_labels[n_labeled_samples:] = -1
+mapping = {0: 'A', 1: 'B', 2: 'C', -1: '-1'}
+y_train_missing_strings = np.vectorize(mapping.get)(
+    y_train_missing_labels).astype(object)
+y_train_missing_strings[y_train_missing_labels == -1] = -1
+
+lb = LabelBinarizer()
+y_train_missing_dummies = lb.fit_transform(y_train_missing_labels)
+
+
+def test_missing_predict_proba():
+    # Check that an error is thrown if predict_proba is not implemented
+    base_classifier = SVC(gamma="scale")
+    self_training = SelfTrainingClassifier(base_classifier)
+    message = "base_classifier (SVC) should implement predict_proba!"
+    assert_raise_message(ValueError, message, self_training.fit, X_train,
+                         y_train)
+
+
+def test_none_classifier():
+    st = SelfTrainingClassifier(None)
+    msg = "base_classifier cannot be None"
+    assert_raise_message(ValueError, msg, st.fit, X_train,
+                         y_train_missing_labels)
+
+
+@pytest.mark.parametrize("max_iter, threshold",
+                         [(-1, 1.0), (-100, -2), (-10, 10)])
+def test_invalid_params(max_iter, threshold):
+    # Test negative iterations
+    base_classifier = SVC(gamma="scale", probability=True)
+    st = SelfTrainingClassifier(base_classifier, max_iter=max_iter)
+    message = "max_iter must be >= 0 or None, got"
+    assert_raise_message(ValueError, message, st.fit, X_train, y_train)
+
+    base_classifier = SVC(gamma="scale", probability=True)
+    st = SelfTrainingClassifier(base_classifier, threshold=threshold)
+    message = "threshold must be in [0,1)"
+    assert_raise_message(ValueError, message, st.fit, X_train, y_train)
+
+
+@pytest.mark.parametrize("base_classifier",
```

Comment:
It is entirely unnecessary to test all these parameter combinations. We know that these estimators support the predict_proba API. We only need to check, roughly, that given such an API we can get results. We certainly don't need to check it for each max_iter and threshold. It would be better to test that these parameters *change* the result than to check that they are validly accepted.